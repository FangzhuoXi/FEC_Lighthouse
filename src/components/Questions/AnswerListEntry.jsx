import React, { useState, useEffect } from 'react';
import axios from 'axios';
import moment from 'moment';
import styled from 'styled-components';

const Span = styled.span`
color: #f26938;
text-decoration: underline;

&:hover {
  color: #68a69b;
}
`;

const Answer = styled.h4`
color: #010a26;
`;

const AnswerListEntry = ({answer}) => {

  return (
    <>
      <Answer className="answer_title">{answer.body}</Answer>
      <p className="answer helpful">Helpful? <Span>Yes&#40;#&#41;</Span> <Span>Report</Span></p>
      <h5>{`by: ${answer.answerer_name} ${moment(answer.date).format('MMMM DD, YYYY')}`}</h5>
    </>
  )
}

export default AnswerListEntry