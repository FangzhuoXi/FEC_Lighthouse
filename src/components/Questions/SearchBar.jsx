import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { SearchBarInput, Button } from './Styles.jsx';

const SearchBar = ({ searchQuestions }) => {

  const [search, setSearch] = useState('');

  const searchInput = (query) => {
    setSearch(query);
    if (query.length > 2) {
      searchQuestions(query);
    } else {
      searchQuestions();
    }
  }

  const officiallySearch = (e) => {
    e.preventDefault();
    searchQuestions(search);
  }

  return (
    <>
      <SearchBarInput type="text" id="search" name="search" placeholder="Have a question? Search for the answersâ€¦" onChange={e => searchInput(e.target.value)}/>
      <Button onClick={officiallySearch}>&#128269;</Button>
    </>
  )
}

export default SearchBar