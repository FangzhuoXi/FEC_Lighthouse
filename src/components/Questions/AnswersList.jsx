import React, { useState } from 'react';
import axios from 'axios';
import AnswerListEntry from './AnswerListEntry.jsx';
import { AnswerTitle, Button, LeftCol, RightCol,
  AnswerScrollBar } from './Styles.jsx';

const AnswersList = ({ answers, length, allVisible, showAll }) => {

  return (
    <div>
      <LeftCol><AnswerTitle>A:</AnswerTitle></LeftCol>
      <RightCol>
        <AnswerScrollBar>
          {answers.map((answer) => {
            return (
              <AnswerListEntry key={answer.answer_id} answer={answer} />
            )
          })}
          {length > 2 && (allVisible ? <Button onClick={(e) => showAll()}>Collapse Answers</Button> :
            <Button onClick={(e) => showAll()}>See More Answers</Button>)}
        </AnswerScrollBar>
      </RightCol>
    </div>
  )
}

export default AnswersList