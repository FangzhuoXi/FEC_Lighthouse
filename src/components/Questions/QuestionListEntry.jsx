import React, { useEffect, useState } from 'react';
import axios from 'axios';
import moment from 'moment';
import { Input, OuterModal, InnerModal, Span, SpanClicked, EntryTitle, Button, SubTitle, Description } from './Styles.jsx';
import AnswersList from './AnswersList.jsx';
import AddAnswer from './AddAnswerModal.jsx';

const QuestionListEntry = ({question, productName}) => {

  const [answers, setAnswers] = useState([]);
  const [helpfulness, setHelpfulness] = useState(question.question_helpfulness)
  const [addAnswer, setAddAnswer] = useState(false);
  const [isHelpful, setIsHelpful] = useState(false);

  const helpfulClick = (e) => {
    e.preventDefault();
    if (isHelpful === false) {
      setHelpfulness(helpfulness + 1);
      setIsHelpful(true);
      axios.put('/markQuestionHelpful', {params: {question_id: question.question_id}})
        .catch((err) => {
          console.log(`error: ${err}`);
        });

    }
  }

  useEffect(() => {

    axios.get('/getAnswers', {params: {question_id: question.question_id}})
    .then((answerList) => {
      setAnswers(answerList.data.results);
    })
    .catch((err) => {
      console.log(`unable to retrieve answers ${err}`);
    })

  }, []);

  return(
    <>
    <EntryTitle className="question_title">{`Q: ${question.question_body}`}</EntryTitle><SubTitle className="question helpful">Helpful? {isHelpful ? <SpanClicked>{`Yes (${helpfulness})`}</SpanClicked> : <Span onClick={helpfulClick}>{`Yes (${helpfulness})`}</Span>}</SubTitle>
    <Description>{`asked by: ${question.asker_name} ${moment(question.question_date).format('MMMM DD, YYYY')}`}</Description>
    <Button onClick={() => {setAddAnswer(true)}} >Add An Answer</Button>
    <AddAnswer productName={productName} questionBody={question.question_body} questionID={question.question_id} addAnswer={addAnswer} onClose={() => {setAddAnswer(false); return}}/>
    <AnswersList answers={answers}/>
    </>
  )

}

export default QuestionListEntry;