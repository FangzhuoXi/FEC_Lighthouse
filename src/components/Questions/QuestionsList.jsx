import React, { useEffect, useState } from 'react';
import axios from 'axios';
import QuestionListEntry from './QuestionListEntry.jsx';
import { Button, InnerContent, ClearFloat } from './Styles.jsx';

const QuestionsList = ({ productName, questions }) => {

  return (
    <InnerContent>
        <div>
          {questions.map((question) => {
            return (
              <QuestionListEntry key={question.question_id} question={question} productName={productName} />
            )
          })}
        </div>
    </InnerContent>
  )

}

export default QuestionsList;