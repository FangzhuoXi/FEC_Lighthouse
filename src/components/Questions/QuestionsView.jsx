import React, { useEffect, useState } from 'react';
import axios from 'axios';
import SearchBar from './SearchBar.jsx';
import QuestionsList from './QuestionsList.jsx';
import AddQuestion from './AddQuestionModal.jsx';
import AddAnswer from './AddAnswerModal.jsx';

const QuestionsView = ({productId, productName}) => {

  const [questions, setQuestions] = useState([]);

  useEffect(() => {

    axios.get('/getQuestions', {params: {product_id: productId}})
    .then((questionList) => {
      setQuestions(questionList.data.results);
    })
    .catch((err) => {
      console.log(`unable to retrieve questions ${err}`)
    })

  }, []);

  return (
    <>
    <h2>Questions and Answers</h2>
      <SearchBar/>
      <QuestionsList questions={questions}/>
      <button>Add A Question</button>
      <AddQuestion/>
      <AddAnswer/>
    </>
  )
}

export default QuestionsView