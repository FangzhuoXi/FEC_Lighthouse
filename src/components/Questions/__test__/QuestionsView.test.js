import renderer from 'react-test-renderer';
import React, { useEffect, useState } from 'react';
import ReactDOM from "react-dom";

import { render, cleanup } from '@testing-library/react';

import QuestionsView from '../QuestionsView.jsx';
import AddAnswer from '../AddAnswerModal.jsx';
import AddQuestion from '../AddQuestionModal.jsx';
import AnswerListEntry from '../AnswerListEntry.jsx';
import AnswersList from '../AnswersList.jsx';
import QuestionListEntry from '../QuestionListEntry.jsx';
import QuestionsList from '../QuestionsList.jsx';
import SearchBar from '../SearchBar.jsx';

const products = {
  productId: 37311,
  productName: "Camo Onesie",
  question_body: "Why wear a onesie?",
};

const questions = [{  "question_id": 640994,
"question_body": "Where have all the cowboys gone?",
"question_date": "2022-05-25T00:00:00.000Z",
"asker_name": "Someone",
"question_helpfulness": 34,
"reported": false,
"answers": {
    "5985683": {
        "id": 5985683,
        "body": "afsf",
        "date": "2022-05-29T00:00:00.000Z",
        "answerer_name": "cdsv",
        "helpfulness": 1,
        "photos": []
    }
  }
}];

const answers = [{
  "answer_id": 5985322,
  "body": "All I do is rep BrickSquad all day #SoIcy",
  "date": "2022-05-20T00:00:00.000Z",
  "answerer_name": "Liz Penny",
  "helpfulness": 9,
  "photos": []
},
{
  "answer_id": 5985320,
  "body": "that is not a question",
  "date": "2022-05-20T00:00:00.000Z",
  "answerer_name": "dollabills",
  "helpfulness": 2,
  "photos": []
}];

afterEach(cleanup);

test('returns array filled with questions related to product', () => {
  function updateQuestions() {
    let productQuestions = QuestionsView(product);
    expect(productQuestions.questions.length).toBe(80);
  }
});

test('only two questions rendered on initial page load', () => {
  function updateQuestions() {
    let productQuestions = QuestionsView(product);
    expect(productQuestions.visibleQuestions.length).toBe(2);
  }
});

it('renders Questions component without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<QuestionsView />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders submit answer modal without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<AddAnswer />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders submit question modal without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<AddQuestion />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders submit individual answer without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<AnswerListEntry answer={answers[0]}/>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders submit answer list without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<AnswersList answers={answers}/>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders submit individual question without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<QuestionListEntry question={questions[0]}/>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders submit question list without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<QuestionsList questions={questions}/>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders submit question list without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.createRoot(<SearchBar/>, div);
  ReactDOM.unmountComponentAtNode(div);
});

// it('renders AddAnswer modal correctly', () => {
  //   const {getByElement} = render(<AddAnswer productName={product.productName}></AddAnswer>)
  //   expect(getByElement('AddAnswer')).toHave
  // });

  it('matches questions component snapshot',  () => {
    const tree = renderer.create(<QuestionsView></QuestionsView>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches addanswer snapshot', () => {
    const tree = renderer.create(<AddAnswer></AddAnswer>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches addquestion snapshot', () => {
    const tree = renderer.create(<AddQuestion></AddQuestion>);
    expect(tree).toMatchSnapshot();
  });
