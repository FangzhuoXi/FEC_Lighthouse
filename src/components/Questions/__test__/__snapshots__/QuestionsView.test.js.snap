// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches addanswer snapshot 1`] = `
<div
  className="sc-papXJ eyeRQT"
>
  <div
    className="sc-jqUVSM fZdTSS"
  >
    <div
      className="sc-kDDrLX emtVmb"
    >
      <div
        className="sc-iqcoie kFbfYt"
      >
        <h2
          className="sc-hAZoDl hNuKL"
        >
          Submit Your Answer
        </h2>
      </div>
      <h5
        className="sc-idiyUo eVtZXZ"
      >
        undefined: undefined
      </h5>
      <form>
        <h3
          className="sc-bjUoiL prTFN"
        >
          Your Answer *
        </h3>
        <textarea
          className="sc-evZas dLSVMU"
          onChange={[Function]}
          required={true}
          type="text"
        />
        <h3
          className="sc-bjUoiL prTFN"
        >
          What is your nickname? *
        </h3>
        <input
          className="sc-breuTD hrBoTZ"
          onChange={[Function]}
          placeholder="Example: jack543!"
          required={true}
          type="text"
        />
        <p
          className="sc-dIouRR hxmtMU"
        >
          For privacy reasons, do not use your full name or email address
        </p>
        <h3
          className="sc-bjUoiL prTFN"
        >
          Email *
        </h3>
        <input
          className="sc-breuTD hrBoTZ"
          onChange={[Function]}
          placeholder="Example: jack@email.com"
          required={true}
          type="email"
        />
        <p
          className="sc-dIouRR hxmtMU"
        >
          For authentication reasons, you will not be emailed
        </p>
        <button
          className="sc-hHLeRK cqhdEl"
          onClick={[Function]}
        >
          SubmitAnswer
        </button>
        <button
          className="sc-dmRaPn hnwOGC"
        >
          Close
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`matches addquestion snapshot 1`] = `
<div
  className="sc-papXJ eyeRQT"
>
  <div
    className="sc-jqUVSM fZdTSS"
  >
    <div
      className="sc-kDDrLX emtVmb"
    >
      <div
        className="sc-iqcoie kFbfYt"
      >
        <h2
          className="sc-hAZoDl hNuKL"
        >
          Ask Your Question
        </h2>
      </div>
      <h5
        className="sc-idiyUo eVtZXZ"
      >
        About the undefined
      </h5>
      <form>
        <h3
          className="sc-bjUoiL prTFN"
        >
          Your Question *
        </h3>
        <textarea
          className="sc-evZas dLSVMU bodytext"
          onChange={[Function]}
          required={true}
          type="text"
        />
        <h3
          className="sc-bjUoiL prTFN"
        >
          What is your nickname? *
        </h3>
        <input
          className="sc-breuTD hrBoTZ"
          onChange={[Function]}
          placeholder="Example: jackson11!"
          required={true}
          type="text"
        />
        <p
          className="sc-dIouRR hxmtMU"
        >
          For privacy reasons, do not use your full name or email address
        </p>
        <h3
          className="sc-bjUoiL prTFN"
        >
          Email *
        </h3>
        <input
          className="sc-breuTD hrBoTZ"
          onChange={[Function]}
          placeholder="Example: jackson@email.com"
          required={true}
          type="email"
        />
        <p
          className="sc-dIouRR hxmtMU"
        >
          For authentication reasons, you will not be emailed
        </p>
        <button
          className="sc-hHLeRK cqhdEl"
          onClick={[Function]}
        >
          Submit Question
        </button>
        <button
          className="sc-dmRaPn hnwOGC"
        >
          Close
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`matches questions component snapshot 1`] = `
<div
  className="sc-bczRLJ gSbPJH"
>
  <div
    className="sc-ksZaOG jLLjsY"
  >
    <h2
      className="sc-hAZoDl hNuKL"
    >
      Questions and Answers
    </h2>
  </div>
  <div
    className="sc-hKMtZM emBaIV"
  >
    <input
      className="sc-crXcEl fqMCSk"
      id="search"
      name="search"
      onChange={[Function]}
      placeholder="Have a question? Search for the answers‚Ä¶"
      type="text"
    />
    <button
      className="sc-dmRaPn hnwOGC"
      onClick={[Function]}
    >
      üîç
    </button>
    <div
      className="sc-eCYdqJ ktwMtI"
    >
      <button
        className="sc-hHLeRK cqhdEl"
        onClick={[Function]}
      >
        Add A Question
      </button>
    </div>
  </div>
</div>
`;
